09-intermediate-rails/rails-api-intro.md

```diff
@@ -0,0 +1,189 @@
+  </summary>
+
+- Web apps produce **HTML** for humans to consume
+- APIs produce **JSON** for machines to consume
+</details>
+
+What are some of the key similarities? They both use **routes** to provide structure, and there is also some sort of **data** involved. Many APIs (including the one we'll build today) provide some sort of CRUD functionality, and our RESTful design principals and general Rails knowledge will continue to serve us well.
+
+## Creating an API
+
+We want to build a small Rails app that will act as an API for providing data about our amazing, adorable pets to other applications. The app will have the following routes:
+
+- `GET /pets` show all pets
+- `GET /pets/:id` shows a pet with the provided id
+- `POST /pets` add a pet to the collection
+
+Given the context of our application, we should have a model and controller that reference our main resource, pets. Once you clone this repo, you'll notice that we have these things already created for you!
+
+[https://github.com/AdaGold/ada-pets](https://github.com/AdaGold/ada-pets)
+
+To make it go:
+
+```bash
+$ git clone https://github.com/AdaGold/ada-pets.git
+$ cd ada-pets
+$ bundle install
+$ rails db:reset
+$ rails server
+```
+
+Once you've downloaded it, take a few minutes to go through this Rails app with the person next to you.
+
+### Questions
+
+- What components does the project have?
+- What routes are in the project currently?
+  - Do they work?
+- What steps would you perform to set up a project like this from scratch?
+
+## Listing Pets
+
+We are going to approach this application creation from a TDD approach. First we will create a test and then write controller methods to make them pass. You can run the tests using the usual `rails test`.
+
+You'll notice that we already have one basic get route created for pets, and one basic controller action that corresponds with that route.
+
+Try go to to `/pets` route.  What happens?
+
+The terminal should show something like:
+
+```bash
+Completed 204 No Content in 4ms (ActiveRecord: 0.0ms)
+```
+
+This is because an API application does not have a standard view template and our controller does not tell Rails what content to render back to the user.
+
+```ruby
+  def index
+    @pets = Pet.all
+  end
+```
+
+So lets write a test to specify what we **want** to be rendered back to us on `get '/pets'`.
+
+### So... What Do We Want
+
+With your neighbor answer the following questions:
+
+1. What kind of information should be returned here?  HTML?
+1. How much information should be returned?
+1. What information about each pet should be included?
+1. What response code should be returned?  Should it ever vary?
+
+<details>
+  <summary>Our Answer</summary>
+  In this we will expect:
+
+- JSON Data to be returned and not HTML
+- All of our fixture data returned in an array
+- We will make the executive decision to only include the:
+  - id
+  - name
+  - age
+  - human
+
+Which means we will need to exclude some fields here.
```

Comment:
I just think this could be helpful because I think some student might not have thought of those fields and be wondering "what are these other fields a pet has that we're excluding?"